### ISSUE TITLE
Socket.io Client doesn't (re)emit without page reload

### ISSUE THREAD

**Issue by T3KN0KRAT** [socketio/socket.io-client#700]
TL;DR; I can connect to the node.js server and retrieve my data when I click on `<button>Get Data</button>`.  But nothing is emitted if I click on that button again.

I'm not sure if this is a bug, but using socket.io 1.0.4 . . .

... on document ready, my client connects to the node socket.io server to get the client .js file:

``` javascript
    /*** Attempt to get the socket.io.js file from the node listener. ***/
    var svrPort = 8072
    var scriptPath = "https://"+location.host+":"+svrPort+'/socket.io/socket.io.js';
    jQuery.getScript( scriptPath, function( data, textStatus, jqxhr ) {
        if (jqxhr.status == 200) {
            listenerUp = true;
            console.debug( "socket.io.js file received." );
        }
    }); 
```

The page has a `<button id='RS_Refresh'>Get Data</button>` that has an event trigger defined at:

``` javascript
jQuery('#RS_Refresh').click(function(){
   get_AllRS(true); 
});
```

The `get_ALLRS()` function does indeed establish a connection to the node server, and on connect, emits a request for the data to be displayed on the web page, which works perfectly:

``` javascript
/* Some code omitted for brevity */
function get_AllRS(showSpinner) {
    var getAllRS_sock = io.connect(svr,options);
    var toListener =
    {
        'cType'     : 'get_AllRS',
        'DEBUG'     : RS_DEBUG,
        'clientID'  : clientID,
        'userID'    : userID,
    };

   /* On second attempt, this gets executed, but no request is actually 
      emitted  - no traffic seen in Firebug's 'Net' tab */
    if (getAllRS_sock.connected) {
        console.log('emitting. . .');
        getAllRS_sock.emit('get_AllRS',toListener) ;
    }

    /*** When connected, send get_AllRS ***/
    getAllRS_sock.on('connect', function getAllRS_sock_on_connect() {
        console.log('In getAllRS_sock_on_connect()');

        /*** Let's talk to the server ***/
        getAllRS_sock.emit('get_AllRS',toListener) ;
    });

    /*** Mission Accomplished ***/
    getAllRS_sock.on('got_AllRS', function getAllRS_sock_on_got_AllRS(incoming) {
        console.log('In getAllRS_sock_on_got_AllRS()');

        console.debug('=== got_AllRS incoming ======================');
        console.debug(incoming)
        console.debug('=============================================');

        if (typeof(incoming) !== 'undefined'){
            populateRSFields(incoming);
            setRSAlert('Fields Retrieved successfully','flash_success');
        }
        else{
            revertValues();
            setRSAlert('Fields were NOT retrieved','flash_error');
        }

        /* Should I disconnect from the node server?? Doesn't work either way */ 
        // getAllRS_sock.emit('forceDisconnect',{}) ;
        // getAllRS_sock.destroy();
    });
}
```

However, if I click on the _Get Data_ button again, nothing is emitted, according to the 'Net' tab in Firebug and node.js doesn't hear/see the second request in DEBUG=\* mode, ... until I refresh the entire web page.

What gives?  

Should I emit to the server to disconnect the client (i.e. `.emit('forceDisconnect',{}`)? (The problem seems to persist either way, unless I'm doing it wrong.)

Is my approach incorrect?  Or maybe you could show me an example of how you made this work?

**Comment by PgPedroGabriel** [comment]
Thank you @t3knos,  u solve my problem talking about multiplex.
My problem it was i dont can reconnect when i use 'socket.disconnect()' in other times in my aplication using angular

**Comment by tatsujin1** [comment]
I think I have the exact same problem, but your solution didn't work for me :(

I enabled debugging in the client and it seems that the `emit()` stops half-way in the client...

Normal emit looks like this:
```
socket.io-client:socket emitting packet with ack id 0 +27ms
socket.io-1.3.7.min.js:1 socket.io-client:manager writing packet %j +29ms 
Object {type: 2, data: Array[1], id: 0, nsp: "/"}
socket.io-1.3.7.min.js:1 socket.io-parser encoding packet %j +5ms 
Object {type: 2, data: Array[1], id: 0, nsp: "/"}
```
followed by more message-specific stuff...

Then, seemingly random, it stops working. Then only the first message is output to the console ("emitting packet").

I've not been able to figure out when this is happening.
Reconnecting, without reloading the page, works fine, connection established. But the emits still won't work...

**Comment by tatsujin1** [comment]
Any progress on this?

The problem still exists in 1.4.5.

As a workaround, using a timeout would be usable (or slightly less unusable), i.e. if reply isn't received within a few seconds, reconnect and try again. But it seems that socket.io doesn't have timeout nor reconnect facilities, at least not exposed API.

**Comment by paduc** [comment]
Ok i've been debugging a similar problem (first emit was OK, next emits failed).
The problem came from my server-side listener. There was an error in the handler that was totally silent (probably caught somehow by socket.io). This error caused my emits to fail. Fixing the handler fixed the problem.

**Comment by tatsujin1** [comment]
Hm, interesting. Can you give details about what kind of problem it was?
Did that problem result in the same symptoms on the client side that emits never gets sent?

**Comment by paduc** [comment]
The problem was simply a undefined pointer (my handler had two arguments and the code used both, whereas it was being called only with one).
The symptoms were simple: the first emit() would work, the following emit() resulted in no action whatsoever from the server side. There was no indication of a problem on the client or the server sides.

**Comment by tatsujin1** [comment]
Ok thanks for the details I'll go over which emit is sent before it stops working.
However, in my case it only sometimes stops working. Even though the message sequence at page load is always the same.

**Comment by Haideralee** [comment]
I am facing the same issue but in my case the problem in `on` section, emit is working fine but the listen is not working as expected.
Ex: 
```
Ctrl 1
       this.SocketIo.socket.on('user:updated', function(data){
            console.log("user:updated in choose place", data)
        });

Ctrl 2

       this.SocketIo.socket.on('user:updated', function(data){
            console.log("user:updated in choose place", data)
        });

        this.SocketIo.socket.emit('acceptedJob', {res});

Server

     socket.on('acceptedJob', (res) => {
            socket.log('acceptedJob');
            socket.emit('user:updated', res);
      });
```
the confusing thing is on running in `Ctrl 2` but not listen in `Ctrl 1`

**Comment by darrachequesne** [comment]
@Haideralee please make sure the listener is properly initialized in your `Ctrl 1`.

Closing, as the initial issue should be fixed by now.

**Comment by webb24h** [comment]
I confirm this problem still persists in 2021. Using socket.io@4.0.1.
